### PR TITLE
Update Dockerfile (GPU Support, Workdir, Permissions)

### PR DESCRIPTION
**Pull request by ffalkenberg**
This PR updates the Dockerfile to leverage NVIDIA's TensorFlow container, ensuring better GPU support, changes the sources and creates a workdir.

**Changes:**

1. **Base Image Update:** 
   - **Old:** `FROM python:3.8-slim`
   - **New:** `FROM nvcr.io/nvidia/tensorflow:22.10.1-tf2-py3`
   
   The base image has been updated to use NVIDIA's TensorFlow container. This change ensures that the container is optimized for GPU usage and includes all necessary dependencies for TensorFlow to utilize NVIDIA GPUs.

2. **Environment Variables:** 
   - Added `ENV DOCTR_CACHE_DIR=/app/.cache` to specify a cache directory.
   - Added `ENV PATH=/app/venv/bin:$PATH` to include the virtual environment's bin directory in the PATH.

3. **Working Directory:** 
   - **New:** `WORKDIR /app`
   
   The working directory has been set to `/app` to provide a consistent location for application files.

4. **File Copying:** 
   - **Old:** Files were copied to `/tmp`.
   - **New:** All files are now copied directly to the `/app` directory with `COPY . .`.

5. **Permissions:** 
   - **New:** `chmod -R a+w /app`
   
   Permissions have been updated to ensure that all users have write access to the `/app` directory. This change is crucial as some environments, like Kubernetes/OpenShift, do not allow containers to run as root. By adjusting the permissions, we ensure compatibility with such environments.

**Additional Information:**

- **General-Purpose Usage:** The updated Dockerfile is more versatile due to the inclusion of additional sources. It's designed to be general-purpose, catering to a broader range of applications and supporting both CPU and GPU environments.

- **CUDA & cuDNN Dependency:** The new base image from NVIDIA's TensorFlow container comes with CUDA and cuDNN pre-installed, ensuring that TensorFlow can leverage GPU acceleration out of the box. More details can be found on TensorFlow's official [documentation](https://www.tensorflow.org/install/source#gpu) and NVIDIA's [container catalog](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tensorflow/tags).

- **Demo App:** With the inclusion of more sources in the Dockerfile, running applications like the demo app becomes more straightforward and optimized.

### CODE REVIEW

**Comment by ffalkenberg** [comment]
@felixdittrich92 can you add a reviewer?

**Comment by felixdittrich92** [comment]
Hi @ffalkenberg :wave:,

Thanks a lot for the PR.

Added @odulcy-mindee because he works currently on the requested docker features :)

File: Dockerfile [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
+ENV DOCTR_CACHE_DIR=/app/.cache 
+ENV PATH=/app/venv/bin:$PATH
 
+WORKDIR /app
 
-COPY ./pyproject.toml /tmp/pyproject.toml
-COPY ./README.md /tmp/README.md
-COPY ./setup.py /tmp/setup.py
-COPY ./doctr /tmp/doctr
+COPY . .
```
Comment (felixT2K) [inline Dockerfile:10]: @ffalkenberg what's the reason to copy everything ?
I think we should exclude parts like the docs folder, and so on :)

Comment (ffalkenberg) [reply Dockerfile:10]: @felixT2K Thank you for pointing this out.

Here's the breakdown from the `du` output for the `doctr` directory:

```
57M     doctr/.git
100K    doctr/.github
92K     doctr/api
28K     doctr/demo
4.0M    doctr/docs
1.3M    doctr/doctr
8.0K    doctr/notebooks
268K    doctr/references
52K     doctr/scripts
348K    doctr/tests
63M     doctr/
```

From the output, we can observe that while certain directories like `doctr/docs` have a relatively larger size (4.0M) compared to others, the potential memory savings by excluding them would still be negligible in my opionion, but that would be your choice. 

The inclusion of these directories was a simpler and more direct approach, and I specifically wanted to retain directories such as `demo` and `api` based on my needs.

**Comment by odulcy-mindee** [comment]
Hey @ffalkenberg ! Thank you for the PR! 👍
I'll have a look at it !

File: Dockerfile [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
-    && pip install --upgrade pip setuptools wheel \
-    && pip install -e /tmp/.[tf] \
-    && pip cache purge \
+    && apt-get install --no-install-recommends ffmpeg libsm6 libxext6 python3-venv -y \
     && apt-get autoremove -y \
-    && rm -rf /var/lib/apt/lists/* \
-    && rm -rf /root/.cache/pip
+    && rm -rf /var/lib/apt/lists/* 
+
+RUN python -m venv /app/venv \
```
[the remark this answers is not in the document]
Comment (felixT2K) [reply Dockerfile:17]: agree

**Comment by felixT2K** [comment]
@odulcy-mindee 
But if we create a job for docker hub (as discussed internally) do we still need this single dockerfile ??
Otherwise we should have 4 (tf-gpu / pt-gpu / tf-cpu / pt-cpu) ??

File: Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ffalkenberg) [reply Dockerfile:17]: @odulcy-mindee Thank you for your insights on the use of virtual environments within Docker. I introduced the `venv` primarily to address a specific challenge I encountered. During runtime, I needed to install additional packages inside the container, such as Streamlit for demonstrations. Without the virtual environment, I ran into permission issues due to not being root.

If there's an alternative approach that allows for this flexibility without the need for a `venv`, I'm absolutely open to it. I'm always in favor of simplifying our setup while ensuring functionality

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,20 @@
-FROM python:3.8-slim
+FROM nvcr.io/nvidia/tensorflow:22.10.1-tf2-py3
```
[the remark this answers is not in the document]
Comment (ffalkenberg) [reply Dockerfile:1]: @odulcy-mindee, the image does seem oversized. If the tensorflow-gpu image meets our needs (gpu workload using cuda), it's preferable. 
From my experience, when crafting your own image though, starting with a CUDA image is key.

**Comment by odulcy-mindee** [comment]
@felixT2K yeah, good question. It'll depends if we can fit everything in one Dockerfile and then use `--build-arg` to customize it for each version.

Personally, I'd like to have only one Dockerfile that the CI uses but still usable by the users so they can build the docker container locally if needed

**Comment by felixT2K** [comment]
> @felixT2K yeah, good question. It'll depends if we can fit everything in one Dockerfile and then use `--build-arg` to customize it for each version.
> 
> Personally, I'd like to have only one Dockerfile that the CI uses but still usable by the users so they can build the docker container locally if needed

This sounds much better for me @ffalkenberg would it be an option ?
Because we have planned to publish multiple images on docker hub so it would be good to have one flexible dockerfile which can be used in the CI to push multible versions (cpu | gpu / tf | pt / py3.8 | py3.9 | py3.10)

**Comment by odulcy-mindee** [comment]
We can take some inspiration from https://github.com/tensorflow/build/tree/master/tensorflow_runtime_dockerfiles for TensorFlow. I'm looking for a similar resource for PyTorch.

**Comment by felixT2K** [comment]
> We can take some inspiration from https://github.com/tensorflow/build/tree/master/tensorflow_runtime_dockerfiles for TensorFlow. I'm looking for a similar resource for PyTorch.

https://github.com/cnstark/pytorch-docker/blob/main/docker/ubuntu/Dockerfile  with some smaller modifications ?

**Comment by odulcy-mindee** [comment]
@felixT2K Ah, good idea! :+1:

**Comment by ffalkenberg** [comment]
yes that sounds good to me, you can push changes to my branch or close this and create a new branch? Either way is fine for me😊

**Comment by felixT2K** [comment]
> yes that sounds good to me, you can push changes to my branch or close this and create a new branch? Either way is fine for me😊

Do you not want to work on it ? 🤗

**Comment by felixT2K** [comment]
> sure, as soon as i have some free time

alright no stress :)

**Comment by ffalkenberg** [comment]
@odulcy-mindee @felixT2K 

I've now made these changes:

1. **Switched to TensorFlow GPU Base Image**: I've updated our base image to use the TensorFlow GPU version. This should provide better performance for those with GPU setups, but it's also compatible with CPU-only environments.

2. **Removed the `venv` Approach**: As you mentioned this would work in most usecases. I've made the necessary adjustments to my private deployment.yaml to make additional installs as non-root possible.

3. **Single Dockerfile for the Project**: In my opinion, having multiple Dockerfiles for this project might overcomplicate things. The updated Dockerfile is designed to work seamlessly in both CPU and GPU environments.

Please review the changes and let me know if there are any concerns or suggestions. Thanks!

**Comment by felixdittrich92** [comment]
Hi @ffalkenberg 👋,
Thanks for the update.
We should make it possible to specify the base image and python version via build args because the idea is to use it also for pytorch (TF: CPU/GPU | PT: CPU/GPU) / different python versions :)

Afterwards we can add a CI job to push prebuild images for TF/PT to docker hub (in another PR)

**Comment by ffalkenberg** [comment]
@felixdittrich92 Your proposal certainly introduces broader flexibility to the Docker image configuration, but I believe it extends beyond the scope of this pull request. My concerns are:

1. **Complexity and Maintenance**: Adopting different base images introduces variability due to their unique libraries and tools. This can result in potential inconsistencies and, undoubtedly, a higher testing and maintenance burden.
2. **Purpose and Clarity**: I'm not entirely certain of the motivation behind such extensive flexibility. While I acknowledge its potential advantages, I'd suggest considering this enhancement in a separate pull request.

Given your approach, it might be more fitting to construct the image from scratch, integrating elements like CUDA, Python, Torch, etc. However, juxtaposing this with the simplicity of the previous Dockerfile, I'm unsure if this complexity is warranted.

To strike a balance, I can offer to rename the current Dockerfile to `gpu.Dockerfile`, allowing it to coexist with the original one. This provides a clear distinction for now, and a more comprehensive overhaul can be contemplated by your team at a later stage.

**Comment by odulcy-mindee** [comment]
@felixdittrich92 I'm fine to apply this modification to this file, it's good enough.

I'm working on a more generic Dockerfile in https://github.com/mindee/doctr/pull/1322.

**Comment by felixdittrich92** [comment]
> @felixdittrich92 I'm fine to apply this modification to this file, it's good enough.
> 
> I'm working on a more generic Dockerfile in #1322.

Fine for me it's your topic/task ^^

**Comment by felixdittrich92** [comment]
@odulcy-mindee you can merge this if you are fine with the changes.
I have only one question if we build more flexible docker files in #1322 whats the reason to keep this docker file in root ?
Personally i have had also in mind to create a `dockerfiles` folder to store the more generic ones (one TF and one PT) which would make the dockerfile in root obsolete

**Comment by odulcy-mindee** [comment]
> @odulcy-mindee you can merge this if you are fine with the changes. I have only one question if we build more flexible docker files in #1322 whats the reason to keep this docker file in root ? Personally i have had also in mind to create a `dockerfiles` folder to store the more generic ones (one TF and one PT) which would make the dockerfile in root obsolete

Yeah, I totally agree with you. I'm still not entirely sure whether there will be just one file, but I'm working on it. Maybe one file, or a folder `dockerfiles` :man_shrugging:

So maybe this Dockerfile will be obsolote in few days but I'm fine to release this modification if it can help people right now.

**Pull request merged**